File: src/dataforge/core/schema.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class ForeignKey:
    ref_table: str
    ref_column: str


@dataclass
class Column:
    name: str
    dtype: str  # "int", "str", "float", "date", "bool", "email", "uuid", "int_seq"
    faker_provider: str | None = None
    nullable: float = 0.0
    primary_key: bool = False
    foreign_key: ForeignKey | None = None
    min_value: str | int | float | None = None  # for int/float: numeric; for date: "YYYY-MM-DD"
    max_value: str | int | float | None = None  # for int/float: numeric; for date: "YYYY-MM-DD"
    choices: list[str] | None = (
        None  # fixed set of values to sample from (overrides dtype/faker_provider)
    )


@dataclass
class Table:
    name: str
    columns: list[Column]
    default_rows: int = 1000


@dataclass
class DomainSchema:
    name: str
    tables: list[Table] = field(default_factory=list)

File: src/dataforge/frontend/src/types/schema.ts
export type Column = {
  id: string;
  name: string;
  dtype: string;
  isPrimaryKey: boolean;
  fakerProvider: string;
  nullable: string;
  isForeignKey: boolean;
  fkTable: string;
  fkColumn: string;
  min: string;
  max: string;
  choices: string[];
};

export type Table = {
  id: string;
  name: string;
  rows: number;
  columns: Column[];
  position?: { x: number; y: number };
};

export type Schema = {
  domain: string;
  tables: Table[];
};
